### PR TITLE
add check if attribute value is possible to be set as configurable option

### PR DESCRIPTION
**Pull request by brosenberger**
a none existing value would be set and the dropdown would show an empty value (not the please select message)

### Description (*)

Configurable options of a product can be preselected when adding url-parameters for each option (http://shop.sample.com/product-url#<<option1>>=<<option-value1>>). if the option value does not exist, the value would still be set for the option. the result is than an empty dropdown for this option.

with adaption of this value setting mechanism to additionally check if the option value for the option is allowed the wrongly set value is ignored.

### Manual testing scenarios (*)

1. go to a product detail page of a configurable product
2. get the option id and one option value from the html
3. add the option id and value to the url (with ...#<<optionid>>=<<optionvalue>>
4. with an existing option value the according option value should be shown in the product detail page
5. with a non existing option value the according option should just show the please select option

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by brosenberger** [comment]
@DrewML could you give me some hints why the static tests failed? (i only see skipped tests)

**Comment by krzksz** [comment]
@brosenberger Hey, this sometimes happen when there are more then a few errors, in such cases it is best to run the tests locally to check what's going on. Here's `jscs` output:
```
requireOperatorBeforeLineBreak: Operator && should not be on a new line at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   140 |
   141 |            $.each(queryParams, $.proxy(function (key, value) {
   142 |                if (this.options.spConfig.attributes[key] !== undefined
-------------------------------------------------------------------------------^
   143 |                    && this.options.spConfig.attributes[key].options.find(function(element) {
   144 |                        return element.id===value;

requireSpaceAfterKeywords: One space required after "function" keyword at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   141 |            $.each(queryParams, $.proxy(function (key, value) {
   142 |                if (this.options.spConfig.attributes[key] !== undefined
   143 |                    && this.options.spConfig.attributes[key].options.find(function(element) {
------------------------------------------------------------------------------------------^
   144 |                        return element.id===value;
   145 |                    })) {

requireSpacesInAnonymousFunctionExpression: Missing space before opening round brace at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   141 |            $.each(queryParams, $.proxy(function (key, value) {
   142 |                if (this.options.spConfig.attributes[key] !== undefined
   143 |                    && this.options.spConfig.attributes[key].options.find(function(element) {
------------------------------------------------------------------------------------------^
   144 |                        return element.id===value;
   145 |                    })) {

requireSpaceBeforeBinaryOperators: Operator === should not stick to preceding expression at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   142 |                if (this.options.spConfig.attributes[key] !== undefined
   143 |                    && this.options.spConfig.attributes[key].options.find(function(element) {
   144 |                        return element.id===value;
-------------------------------------------------^
   145 |                    })) {
   146 |                    this.options.values[key] = value;

requireSpaceAfterBinaryOperators: Operator === should not stick to following expression at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   142 |                if (this.options.spConfig.attributes[key] !== undefined
   143 |                    && this.options.spConfig.attributes[key].options.find(function(element) {
   144 |                        return element.id===value;
----------------------------------------------------^
   145 |                    })) {
   146 |                    this.options.values[key] = value;

requireOperatorBeforeLineBreak: Operator && should not be on a new line at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   162 |                    attributeId = element.id.replace(/[a-z]*/, '');
   163 |
   164 |                    if (this.options.spConfig.attributes[attributeId] !== undefined
-------------------------------------------------------------------------------------------^
   165 |                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
   166 |                            return optionElement.id===element.value;

requireSpaceAfterKeywords: One space required after "function" keyword at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   163 |
   164 |                    if (this.options.spConfig.attributes[attributeId] !== undefined
   165 |                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
------------------------------------------------------------------------------------------------------^
   166 |                            return optionElement.id===element.value;
   167 |                        })) {

requireSpacesInAnonymousFunctionExpression: Missing space before opening round brace at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   163 |
   164 |                    if (this.options.spConfig.attributes[attributeId] !== undefined
   165 |                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
------------------------------------------------------------------------------------------------------^
   166 |                            return optionElement.id===element.value;
   167 |                        })) {

requireSpaceBeforeBinaryOperators: Operator === should not stick to preceding expression at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   164 |                    if (this.options.spConfig.attributes[attributeId] !== undefined
   165 |                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
   166 |                            return optionElement.id===element.value;
-----------------------------------------------------------^
   167 |                        })) {
   168 |                        this.options.values[attributeId] = element.value;

requireSpaceAfterBinaryOperators: Operator === should not stick to following expression at vendor/magento/module-configurable-product/view/frontend/web/js/configurable.js :
   164 |                    if (this.options.spConfig.attributes[attributeId] !== undefined
   165 |                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
   166 |                            return optionElement.id===element.value;
--------------------------------------------------------------^
   167 |                        })) {
   168 |                        this.options.values[attributeId] = element.value;


10 code style errors found.
```
Eslint seems to not complain about anything.

File: app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js [hunk 1]
```diff
@@ -155,7 +160,13 @@ define([
 
                 if (element.value) {
                     attributeId = element.id.replace(/[a-z]*/, '');
-                    this.options.values[attributeId] = element.value;
+
+                    if (this.options.spConfig.attributes[attributeId] !== undefined
+                        && this.options.spConfig.attributes[attributeId].options.find(function(optionElement) {
```
Comment (krzksz) [inline app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]: Same as above.

**Review by krzksz — CHANGES_REQUESTED** [review]

File: app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js (hunk 1 above)
Comment (brosenberger) [reply app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]: yes you are correct, but i do apply the rule of three (https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) and therefor adapted the second occurrence as paste and copy (for my problem i wouldn't need that change, but added it for consistency)

Comment (brosenberger) [reply app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]: yes you are correct, but i do apply the rule of three (https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) and therefor adapted the second occurrence as paste and copy (for my problem i wouldn't need that change, but added it for consistency)

Comment (krzksz) [reply app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]: We discuss this and still I would like to ask you to make this change. Such comparisons are part of the legacy code and we are trying to refactor them as we go.

**Review by krzksz — APPROVED** [review]
Approving this as the refactor can be done in a separate pull request.

**Comment by magento-engcom-team** [comment]
Hi @krzksz, thank you for the review.
[ENGCOM-6430](https://jira.corp.magento.com/browse/ENGCOM-6430) has been created to process this Pull Request
:eight_spoked_asterisk: @krzksz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
Hi @brosenberger  .
     Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated tests.
Thanks!

**Comment by engcom-Golf** [comment]
HI @krzksz all changes cover by jasmine test could you please review ?

**Comment by brosenberger** [comment]
thanks for the testcoverage @Nazar65

**Comment by engcom-Golf** [comment]
@magento run all tests

**Review by krzksz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @krzksz, thank you for the review.
[ENGCOM-6430](https://jira.corp.magento.com/browse/ENGCOM-6430) has been created to process this Pull Request

File: app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js [hunk 2]
```diff
@@ -139,7 +139,12 @@ define([
             });
 
             $.each(queryParams, $.proxy(function (key, value) {
-                this.options.values[key] = value;
+                if (this.options.spConfig.attributes[key] !== undefined &&
+                    this.options.spConfig.attributes[key].options.find(function (element) {
```
Comment (krzksz) [inline app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]:
```suggestion
                    _.find(this.options.spConfig.attributes[key].options, function (element) {
```

File: app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js [hunk 3]
```diff
@@ -155,7 +160,13 @@ define([
 
                 if (element.value) {
                     attributeId = element.id.replace(/[a-z]*/, '');
-                    this.options.values[attributeId] = element.value;
+
+                    if (this.options.spConfig.attributes[attributeId] !== undefined &&
+                        this.options.spConfig.attributes[attributeId].options.find(function (optionElement) {
```
Comment (krzksz) [inline app/code/Magento/ConfigurableProduct/view/frontend/web/js/configurable.js]:
```suggestion
                        _.find(this.options.spConfig.attributes[attributeId].options, function (optionElement) {
```

File: dev/tests/js/jasmine/tests/app/code/Magento/ConfigurableProduct/view/frontend/js/configurable.test.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            'data-validate=\'{required:true}\'' +
+            'id=\'attribute93\'' +
+            'class=\'super-attribute-select\'>' +
+            '<option value=\'\'></option>' +
+            '</select>',
+            selectElement = $(option);
+
+    beforeEach(function () {
+        //remove this lines when jasmine version will be upgraded
+        if (!Array.prototype.find) {
```
Comment (krzksz) [inline dev/tests/js/jasmine/tests/app/code/Magento/ConfigurableProduct/view/frontend/js/configurable.test.js]: PR would no longer use `Array.prototype.find` so we can remove it.

**Review by krzksz — CHANGES_REQUESTED** [review]
I'm sorry I missed that earlier but using `Array.prototype.find` is not possible as it is not supported by IE which we need to cover in Magento. I added the suggestion how to easily replace it with Underscore's equivalent.

**Review by krzksz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @krzksz, thank you for the review.
[ENGCOM-6430](https://jira.corp.magento.com/browse/ENGCOM-6430) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**